### ISSUE TITLE
noise_distribution and sample_schedule do not multiply sigma_data

### ISSUE THREAD

**Issue by wufandi** [lucidrains/alphafold3-pytorch#124]
noise_distribution
https://github.com/lucidrains/alphafold3-pytorch/blob/8ef24ebdd69b35874e067e78afa0367b2fcfadef/alphafold3_pytorch/alphafold3.py#L2188

and
sample_schedule
https://github.com/lucidrains/alphafold3-pytorch/blob/8ef24ebdd69b35874e067e78afa0367b2fcfadef/alphafold3_pytorch/alphafold3.py#L2106C9-L2106C24

should be multiplied self.sigma_data follow equation 7 in page 24 of SI.

**Comment by lucidrains** [comment]
@wufandi they are taken into account [here](https://github.com/lucidrains/alphafold3-pytorch/blob/8ef24ebdd69b35874e067e78afa0367b2fcfadef/alphafold3_pytorch/alphafold3.py#L2056-L2065) ?

**Comment by wufandi** [comment]
I think these are for Algorithm 20 line 2 and Algorithm 20 line 8. This is not the same thing as noise_distribution and sample_schedule.

**Comment by wufandi** [comment]
BTW, step scale in Algorithm 18 is missing in 
https://github.com/lucidrains/alphafold3-pytorch/blob/8ef24ebdd69b35874e067e78afa0367b2fcfadef/alphafold3_pytorch/alphafold3.py#L2167

**Comment by lucidrains** [comment]
@wufandi ok, i'll take a look later 👍 

i'm using a working implementation of EDM for text to image, but maybe they deviated from the original paper

**Comment by lucidrains** [comment]
![347439129-dae8056c-0e45-40bc-b24d-9199a8f3baa7](https://github.com/user-attachments/assets/39a2b8a4-aa83-4e33-b278-03bd09e28529)

here it is from the original paper

**Comment by wufandi** [comment]
Yes, they are different from the original paper. I also agree with the discussion in 

https://github.com/lucidrains/alphafold3-pytorch/issues/68

However, this will make the MSE Loss too small during training. Perhaps the Loss weight of the original paper in AF3 is correct

**Comment by lucidrains** [comment]
@wufandi ok sounds good. the paper is lacking on expounding on these tiny differences.

i have also come across an error in their pseudocode for their diffusion transformer, which is why i'm relying on original work.

but, i can possibly align it with their paper under a flag for the `ElucidatedAtomDiffusion` class. you have started training and it works better with what the paper does?

**Comment by wufandi** [comment]
I also noticed the error you found (residual in AttentionPairBias, Algorithm 23). I tried training with the original version and it seems to converge normally. Maybe adding residual will make the training more stable? I'm not sure.

To verify my code, I chose antibodies in SAbDab as the training/validation set, as I have also done some previous work on antibody structure prediction. In this case, it will work.

**Comment by lucidrains** [comment]
@wufandi hmm, i highly doubt it is stable. take it from someone who knows transformers well

**Comment by lucidrains** [comment]
@wufandi ok, i'll add both the loss weight discrepancy as well as the scaling by sigma data for noise and sample schedules behind a flag

**Comment by wufandi** [comment]
The antibody structure is very similar. so the training is very stable. 
I have added a residual in AttentionPairBias. ;-)

**Comment by lucidrains** [comment]
@wufandi nice, you'll have less headache, guaranteed 😄

**Comment by lucidrains** [comment]
@wufandi ok, let me know if [this](https://github.com/lucidrains/alphafold3-pytorch/commit/1a4506275720b7a1abad6c78ddbd09017d02abb5) looks ok

**Comment by lucidrains** [comment]
> Algorithm 18

yes, you are right, should be added [here](https://github.com/lucidrains/alphafold3-pytorch/commit/b2e492886f6af0770303e4ff4b83d9f76aee096f)

thank you

**Comment by wufandi** [comment]
That looks correct,
One more question, 
https://github.com/lucidrains/alphafold3-pytorch/blob/main/alphafold3_pytorch/alphafold3.py#L2554-L2574 

Why the CentreRandomAugmentation not used during sampling? Follow line 3 in Algorithm 18.

**Comment by lucidrains** [comment]
@wufandi yes, you are right again, thank you for catching that

https://github.com/lucidrains/alphafold3-pytorch/commit/41c58c3f8ce7cd6e3b13eda36fdc2af91ee22e44

**Comment by wufandi** [comment]
AF3 also do not use c_noise, which is used by EDM
https://github.com/lucidrains/alphafold3-pytorch/blob/main/alphafold3_pytorch/alphafold3.py#L2495

**Comment by lucidrains** [comment]
@wufandi ok https://github.com/lucidrains/alphafold3-pytorch/commit/88a65b660aea48ef677299ed08c393bf889c447d

**Comment by lucidrains** [comment]
@wufandi do run the original Karras against the paper's and let me know which one works better

**Comment by lucidrains** [comment]
<img width="901" alt="Screen Shot 2024-08-03 at 11 09 05 AM" src="https://github.com/user-attachments/assets/df3a7037-8bc9-4f79-b5eb-70750c2cf5b3">

oh yes, now i see, they do acknowledge differences, just not well explained

**Comment by lucidrains** [comment]
@wufandi actually, I'll just try the deepmind formulation on generating images tomorrow

**Comment by wufandi** [comment]
I do not run the original Karras EDM, so I am not sure which one works better.

I also find some bug or  in you diffusion module.

1. Algorithm 21 line 8 of AF3 shows that DeepMInd consider the c_noise of EDM paper, you also implemented this algorithm,
https://github.com/lucidrains/alphafold3-pytorch/blob/main/alphafold3_pytorch/alphafold3.py#L1731
so there is not need to add it here,
https://github.com/lucidrains/alphafold3-pytorch/blob/main/alphafold3_pytorch/alphafold3.py#L2531

2. The most important bug in training, which there is no details in the AF3 paper,
Here, it should be
`noised_atom_pos = atom_pos_ground_truth + padded_sigmas * self.c_noise(noise)`
in
https://github.com/lucidrains/alphafold3-pytorch/blob/main/alphafold3_pytorch/alphafold3.py#L2677

In this way, we can ensure that the noise is consistent with the normal distribution.

3. AF3 do not have this process, which is exists in EDM,
https://github.com/lucidrains/alphafold3-pytorch/blob/main/alphafold3_pytorch/alphafold3.py#L2618

**Comment by lucidrains** [comment]
@wufandi thanks Fandi, let's go with your intuition that there may be a bug in the paper. i will continue to keep the formulation from Tero Karras just in case

can you see if the latest version lines up with what you believe should be correct?

**Comment by wufandi** [comment]
Yes, I believe it is correct.

c_noise is really important, otherwise the model will not converge.

**Comment by lucidrains** [comment]
@wufandi aha, i've thought about that possibility after seeing the egregious bug in that transformer residual.. could be 😆

**Comment by wufandi** [comment]
I carefully read the EDM paper (and of course EDM2 as well). I think I made a mistake in the code regarding adding noise to the structure. It should be:
`noised_atom_pos = atom_pos_ground_truth + padded_sigmas * noise`

Please make sure to remove this modification to avoid causing any misunderstanding for the users. However, the part about weighing the loss is most likely an error in the paper.

Additionally, regarding why we need to multiply noise_distribution and sample_schedule by sigma_data, this is because for images, the variance is normalized, whereas for structures, it cannot be normalized.

**Comment by wufandi** [comment]
In addition, after reading the EDM2 paper, I have some extra thoughts:

"The EDM formulation in Equation 4 makes relatively strong assumptions about the mean and standard deviation of the training data. We choose to normalize the training data globally to satisfy these assumptions—as opposed to, e.g., changing the value of sigma_data, which might have far-reaching consequences in terms of the other hyperparameters. We thus keep sigma_data at its default value 0.5, subtract 5.81, 3.25, 0.12, -2.15 from the latents during dataset pre-processing to make them zero mean, and multiply them by 0.5/4.17, 4.62, 3.71, 3.28 to make their standard deviation agree with sigma_data."

This is the trick used in the EDM2 paper. Is it possible for us to set different sigma_data values for different types of molecules?

**Comment by lucidrains** [comment]
> I carefully read the EDM paper (and of course EDM2 as well). I think I made a mistake in the code regarding adding noise to the structure. It should be: `noised_atom_pos = atom_pos_ground_truth + padded_sigmas * noise`
> 
> Please make sure to remove this modification to avoid causing any misunderstanding for the users. However, the part about weighing the loss is most likely an error in the paper.
> 
> Additionally, regarding why we need to multiply noise_distribution and sample_schedule by sigma_data, this is because for images, the variance is normalized, whereas for structures, it cannot be normalized.

sounds good, and no problem :smile:



> In addition, after reading the EDM2 paper, I have some extra thoughts:
> 
> "The EDM formulation in Equation 4 makes relatively strong assumptions about the mean and standard deviation of the training data. We choose to normalize the training data globally to satisfy these assumptions—as opposed to, e.g., changing the value of sigma_data, which might have far-reaching consequences in terms of the other hyperparameters. We thus keep sigma_data at its default value 0.5, subtract 5.81, 3.25, 0.12, -2.15 from the latents during dataset pre-processing to make them zero mean, and multiply them by 0.5/4.17, 4.62, 3.71, 3.28 to make their standard deviation agree with sigma_data."
> 
> This is the trick used in the EDM2 paper. Is it possible for us to set different sigma_data values for different types of molecules?

yup, that may be a possibility

perhaps something like

```python
alphafold3 = Alphafold3(
  ...
  edm_sigma_data = tensor([1., 2., 3., 4., 5.])
)
```

where the indices correspond to the `is_molecule_type` boolean tensor? https://github.com/lucidrains/alphafold3-pytorch/blob/main/alphafold3_pytorch/alphafold3.py#L134

**Comment by lucidrains** [comment]
@wufandi are you referring to https://github.com/lucidrains/alphafold3-pytorch/issues/68 when you say you believe this is an error?
